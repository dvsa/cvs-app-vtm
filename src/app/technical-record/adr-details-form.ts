import {FormGroup} from '@angular/forms';

export class AdrDetailsFormData {
  public static get AdrDetailsForm(): FormGroup {
    return ;
  }
}
