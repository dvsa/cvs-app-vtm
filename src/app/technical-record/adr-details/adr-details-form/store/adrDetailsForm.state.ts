import { createFormGroupState, FormGroupState, box } from 'ngrx-forms';
import { adrDetailsFormModel } from './adrDetailsForm.model';
import { IAppState as RootState } from '../../../../store/state/app.state';

export interface IAppState extends RootState {
  adrDetails: {
    formState: FormGroupState<adrDetailsFormModel>;
    permittedDangerousGoodsOptions: string[];
    additionalNotesOptions: string[];
    productListUnNo: {
      maxIndex: number;
      options: number[];
    };
    tc3Type: {
      maxIndex: number;
      options: number[];
    };
    tc3PeriodicNumber: {
      maxIndex: number;
      options: number[];
    };
    tc3PeriodicExpiryDate: {
      maxIndex: number;
      options: { day: number; month: number; year: number; }[];
    };
    tankDocuments: {
      maxIndex: number;
      options: (string | ArrayBuffer)[];
    };
    submittedValue: adrDetailsFormModel | undefined;
    // msUserDetails: {
    //   msUser: string;
    //   msOid: string;
    // }
  };
}

export const FORM_ID = 'adrDetails';

export const INITIAL_STATE = createFormGroupState<adrDetailsFormModel>(FORM_ID, {
  name: '',
  street: '',
  town: '',
  city: '',
  postcode: '',
  type: '',
  approvalDate: {
    day: 1,
    month: 1,
    year: 1920
  },
  permittedDangerousGoods: {
  },
  compatibilityJ: false,
  additionalNotes: {},
  adrTypeApprovalNo: '',
  tankManufacturer: '',
  yearOfManufacture: 1,
  tankManufacturerSerialNo: '',
  tankTypeAppNo: '',
  tankCode: '',
  tankDocuments: [],
  substancesPermitted: '',
  selectReferenceNumber: '',
  statement: '',
  productListRefNo: '',
  productListUnNo: [],
  productList: '',
  specialProvisions: '',
  tc2Type: '',
  tc2IntermediateApprovalNo: '',
  tc2IntermediateExpiryDate: {
    day: 1,
    month: 1,
    year: 1920
  },
  tc3Type: [],
  tc3PeriodicNumber: [],
  tc3PeriodicExpiryDate: [],
  memosApply: { '07/09 3mth leak ext': false },
  listStatementApplicable: '',
  batteryListNumber: '',
  brakeDeclarationIssuer: '',
  brakeEndurance: '',
  brakeDeclarationsSeen: '',
  declarationsSeen: '',
  weight: 0,
  certificateReq: '',
  adrMoreDetail: ''
});

export function createInitialState(initialAdrDetails: any): any {
  return {
    name: initialAdrDetails.applicantDetails.name,
    street: initialAdrDetails.applicantDetails.street,
    town: initialAdrDetails.applicantDetails.town,
    city: initialAdrDetails.applicantDetails.city,
    postcode: initialAdrDetails.applicantDetails.postcode,
    type: initialAdrDetails.vehicleDetails.type,
    approvalDate: {
      day: initialAdrDetails.vehicleDetails.approvalDate.split('-')[2],
      month: initialAdrDetails.vehicleDetails.approvalDate.split('-')[1],
      year: initialAdrDetails.vehicleDetails.approvalDate.split('-')[0]
    },
    permittedDangerousGoods: {
    },
    compatibilityJ: false,
    additionalNotes: {},
    adrTypeApprovalNo: initialAdrDetails.adrTypeApprovalNo,
    tankManufacturer: initialAdrDetails.hasOwnProperty('tank') ? initialAdrDetails.tank.tankDetails.tankManufacturer : '',
    yearOfManufacture: initialAdrDetails.hasOwnProperty('tank') ? initialAdrDetails.tank.tankDetails.yearOfManufacture : '',
    tankManufacturerSerialNo: initialAdrDetails.hasOwnProperty('tank') ? initialAdrDetails.tank.tankDetails.tankManufacturerSerialNo : '',
    tankTypeAppNo: initialAdrDetails.hasOwnProperty('tank') ? initialAdrDetails.tank.tankDetails.tankTypeAppNo : '',
    tankCode: initialAdrDetails.hasOwnProperty('tank') ? initialAdrDetails.tank.tankDetails.tankCode : '',
    tankDocuments: initialAdrDetails.documents,
    substancesPermitted: initialAdrDetails.hasOwnProperty('tank') ? initialAdrDetails.tank.tankStatement.substancesPermitted : ' ',
    selectReferenceNumber: initialAdrDetails.hasOwnProperty('tank') ?
      (initialAdrDetails.tank.tankStatement ? initialAdrDetails.tank.tankStatement.hasOwnProperty('statement')
        ? 'isStatement' : 'isProductListRefNo' : '') : '',
    statement: initialAdrDetails.hasOwnProperty('tank') ? initialAdrDetails.tank.tankStatement.statement : '',
    productListRefNo: initialAdrDetails.hasOwnProperty('tank') ? initialAdrDetails.tank.tankStatement.productListRefNo : '',
    productListUnNo: [initialAdrDetails.hasOwnProperty('tank') ? initialAdrDetails.tank.tankStatement.productListUnNo : ''],
    productList: initialAdrDetails.hasOwnProperty('tank') ? initialAdrDetails.tank.tankStatement.productList : '',
    specialProvisions: initialAdrDetails.hasOwnProperty('tank') ? initialAdrDetails.tank.tankDetails.specialProvisions : '',
    tc2Type: initialAdrDetails.hasOwnProperty('tank') ? initialAdrDetails.tank.tankDetails.tc2Details.tc2Type : '',
    tc2IntermediateApprovalNo: initialAdrDetails.hasOwnProperty('tank') ? initialAdrDetails.tank.tankDetails.tc2Details.tc2IntermediateApprovalNo : '',
    tc2IntermediateExpiryDate: {
      day: initialAdrDetails.hasOwnProperty('tank') ? (initialAdrDetails.tank.tankDetails.tc2Details.tc2IntermediateExpiryDate
        ? initialAdrDetails.tank.tankDetails.tc2Details.tc2IntermediateExpiryDate.split('-')[2] : '') : '',
      month: initialAdrDetails.hasOwnProperty('tank') ? (initialAdrDetails.tank.tankDetails.tc2Details.tc2IntermediateExpiryDate
        ? initialAdrDetails.tank.tankDetails.tc2Details.tc2IntermediateExpiryDate.split('-')[1] : '') : '',
      year: initialAdrDetails.hasOwnProperty('tank') ? (initialAdrDetails.tank.tankDetails.tc2Details.tc2IntermediateExpiryDate
        ? initialAdrDetails.tank.tankDetails.tc2Details.tc2IntermediateExpiryDate.split('-')[0] : '') : ''
    },
    tc3Type: [],
    tc3PeriodicNumber: [],
    tc3PeriodicExpiryDate: [],
    memosApply: initialAdrDetails.memosApply ? initialAdrDetails.memosApply.includes('07/09 3mth leak ext') ? '07/09 3mth leak ext' : '' : '',
    listStatementApplicable: initialAdrDetails.listStatementApplicable ? 'applicable' : 'notApplicable', // boolean in DB
    batteryListNumber: initialAdrDetails.batteryListNumber,
    brakeDeclarationIssuer: initialAdrDetails.brakeDeclarationIssuer,
    brakeEndurance: initialAdrDetails.brakeEndurance, // boolean in DB
    brakeDeclarationsSeen: initialAdrDetails.brakeDeclarationsSeen,
    declarationsSeen: initialAdrDetails.declarationsSeen, // boolean in DB
    weight: initialAdrDetails.weight,
    certificateReq: initialAdrDetails.additionalNotes.guidanceNotes ? initialAdrDetails.additionalNotes.guidanceNotes.includes('New certificate requested')
    ? 'New certificate requested' : '' : '',
    adrMoreDetail: initialAdrDetails.additionalExaminerNotes
  };
}

export function createSubmitState(adrDetails: any, techRecord: any): any {
  return {
    vehicleDetails: {
      type: adrDetails.type,
      approvalDate: adrDetails.approvalDate.year + '-' + adrDetails.approvalDate.month + '-' + adrDetails.approvalDate.day
    },
    listStatementApplicable: adrDetails.listStatementApplicable === 'applicable',
    batteryListNumber: adrDetails.batteryListNumber,
    declarationSeen: adrDetails.declarationsSeen === 'true', // boolean
    brakeDeclarationsSeen: adrDetails.brakeDeclarationsSeen,
    brakeDeclarationIssuer: adrDetails.brakeDeclarationIssuer,
    brakeEndurance: adrDetails.brakeEndurance, // boolean
    weight: adrDetails.weight,
    compatibilityGroupJ: adrDetails.compatibilityGroupJ,
    permittedDangerousGoods: [
      adrDetails.permittedDangerousGoods
    ],
    additionalExaminerNotes: adrDetails.additionalNotes,
    applicantDetails: {
      name: adrDetails.name,
      street: adrDetails.street,
      town: adrDetails.town,
      city: adrDetails.city,
      postcode: adrDetails.postcode
    },
    memosApply: [
      adrDetails.memosApply
    ],
    additionalNotes: {
      guidanceNotes: [
        adrDetails.guidanceNotes
      ]
    },
    adrTypeApprovalNo: adrDetails.adrTypeApprovalNo,
    tank: {
      tankDetails: {
        tankManufacturer: adrDetails.tankManufacturer,
        yearOfManufacture: adrDetails.yearOfManufacture,
        tankCode: adrDetails.tankCode,
        specialProvisions: adrDetails.specialProvisions,
        tankManufacturerSerialNo: adrDetails.tankManufacturerSerialNo,
        tankTypeAppNo: adrDetails.tankTypeAppNo,
        tc2Details: {
          tc2Type: adrDetails.tc2Type,
          tc2IntermediateApprovalNo: adrDetails.tc2IntermediateApprovalNo,
          tc2IntermediateExpiryDate: adrDetails.tc2IntermediateExpiryDate.year + '-' + adrDetails.tc2IntermediateExpiryDate.month +
          '-' + adrDetails.tc2IntermediateExpiryDate.day
        },
        tc3Details: [
          {
            tc3Type: adrDetails.tc3Type,
            tc3PeriodicNumber: adrDetails.tc3PeriodicNumber,
            tc3PeriodicExpiryDate: adrDetails.tc3PeriodicExpiryDate.year + '-' + adrDetails.tc3PeriodicExpiryDate.month + '-'
            + adrDetails.tc3PeriodicExpiryDate.day
          }
        ]

      },
      tankStatement: {
        substancesPermitted: adrDetails.substancesPermitted,
        statement: adrDetails.statement,
        productListRefNo: adrDetails.productListRefNo,
        productListUnNo: adrDetails.productListUnNo,
        productList: adrDetails.productList
      }
    }
  };
}

