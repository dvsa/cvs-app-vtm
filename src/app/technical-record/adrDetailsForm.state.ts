import { IAppState as RootState } from '@app/store/state/app.state';
import { createFormGroupState, FormGroupState, box } from 'ngrx-forms';
import { adrDetailsFormModel } from './adrDetailsForm.model';

export interface IAppState extends RootState {
  adrDetails: {
    formState: FormGroupState<adrDetailsFormModel>;
    permittedDangerousGoodsOptions: string[];
    additionalNotesOptions: string[];
    productListUnNo: {
      maxIndex: number;
      options: number[];
    };
    tc3Type: {
      maxIndex: number;
      options: number[];
    };
    tc3PeriodicNumber: {
      maxIndex: number;
      options: number[];
    };
    tc3PeriodicExpiryDate: {
      maxIndex: number;
      options: { day: number; month: number; year: number; }[];
    };
    submittedValue: adrDetailsFormModel | undefined;
  };
}

export const FORM_ID = 'adrDetails';

export const INITIAL_STATE = createFormGroupState<adrDetailsFormModel>(FORM_ID, {
  name: '',
  street: '',
  town: '',
  city: '',
  postcode: '',
  type: '',
  approvalDate: {
    day: 1,
    month: 1,
    year: 1920
  },
  permittedDangerousGoods: {
  },
  compatibilityJ: false,
  additionalNotes: {},
  adrTypeApprovalNo: '',
  tankManufacturer: '',
  yearOfManufacture: 1,
  tankManufacturerSerialNo: '',
  tankTypeAppNo: '',
  tankCode: '',
  tankDocuments:{},
  substancesPermitted: '',
  selectReferenceNumber: '',
  statement: '',
  productListRefNo: '',
  productListUnNo: [],
  productList: '',
  specialProvisions: '',
  tc2Type: '',
  tc2IntermediateApprovalNo: '',
  tc2IntermediateExpiryDate: {
    day: 1,
    month: 1,
    year: 1920
  },
  tc3Type: [],
  tc3PeriodicNumber: [],
  tc3PeriodicExpiryDate: [],
  memosApply: { '07/09 3mth leak ext': false },
  listStatementApplicable: '',
  batteryListNumber: '',
  brakeDeclarationIssuer: '',
  brakeEndurance: '',
  brakeDeclarationsSeen: '',
  declarationsSeen: '',
  weight: 0,
  certificateReq: '',
  adrMoreDetail: ''
});
