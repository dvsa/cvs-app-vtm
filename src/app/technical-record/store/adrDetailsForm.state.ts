import { createFormGroupState, FormGroupState, box } from 'ngrx-forms';
import { adrDetailsFormModel } from './adrDetailsForm.model';
import {IAppState as RootState} from '../../store/state/app.state';

export interface IAppState extends RootState {
  adrDetails: {
    formState: FormGroupState<adrDetailsFormModel>;
    permittedDangerousGoodsOptions: string[];
    additionalNotesOptions: string[];
    productListUnNo: {
      maxIndex: number;
      options: number[];
    };
    tc3Type: {
      maxIndex: number;
      options: number[];
    };
    tc3PeriodicNumber: {
      maxIndex: number;
      options: number[];
    };
    tc3PeriodicExpiryDate: {
      maxIndex: number;
      options: { day: number; month: number; year: number; }[];
    };
    tankDocuments: {
      maxIndex: number;
      options: (string | ArrayBuffer) [];
    };
    submittedValue: adrDetailsFormModel | undefined;
  };
}

export const FORM_ID = 'adrDetails';

export const INITIAL_STATE = createFormGroupState<adrDetailsFormModel>(FORM_ID, {
  name: '',
  street: '',
  town: '',
  city: '',
  postcode: '',
  type: '',
  approvalDate: {
    day: 1,
    month: 1,
    year: 1920
  },
  permittedDangerousGoods: {
  },
  compatibilityJ: false,
  additionalNotes: {},
  adrTypeApprovalNo: '',
  tankManufacturer: '',
  yearOfManufacture: 1,
  tankManufacturerSerialNo: '',
  tankTypeAppNo: '',
  tankCode: '',
  tankDocuments:[],
  substancesPermitted: '',
  selectReferenceNumber: '',
  statement: '',
  productListRefNo: '',
  productListUnNo: [],
  productList: '',
  specialProvisions: '',
  tc2Type: '',
  tc2IntermediateApprovalNo: '',
  tc2IntermediateExpiryDate: {
    day: 1,
    month: 1,
    year: 1920
  },
  tc3Type: [],
  tc3PeriodicNumber: [],
  tc3PeriodicExpiryDate: [],
  memosApply: { '07/09 3mth leak ext': false },
  listStatementApplicable: '',
  batteryListNumber: '',
  brakeDeclarationIssuer: '',
  brakeEndurance: '',
  brakeDeclarationsSeen: '',
  declarationsSeen: '',
  weight: 0,
  certificateReq: '',
  adrMoreDetail: ''
});
