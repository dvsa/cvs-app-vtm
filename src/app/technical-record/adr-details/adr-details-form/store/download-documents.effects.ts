import { Injectable } from '@angular/core';
import { Effect, Actions, ofType } from '@ngrx/effects';
import { of } from 'rxjs';
import { catchError, switchMap, tap, withLatestFrom, map } from 'rxjs/operators';
import { TechnicalRecordService } from '@app/technical-record-search/technical-record.service';
import { FileSaverService } from 'ngx-filesaver';
import { Store } from '@ngrx/store';
import { DownloadDocumentFileAction, DownloadDocumentFileActionSuccess, DownloadDocumentFileActionFailure } from './adrDetails.actions';
import { selectVehicleTechRecordModelHavingStatusAll } from '@app/store/selectors/VehicleTechRecordModel.selectors';
import { IAppState } from '@app/store/state/app.state';
import { SetErrorMessage, ClearErrorMessage } from '@app/store/actions/Error.actions';


@Injectable()
export class DownloadDocumentsEffects {
  @Effect()
  downloadDocuments$ = this._actions$.pipe(
    ofType<DownloadDocumentFileAction>(DownloadDocumentFileAction.TYPE),
    withLatestFrom(this._store$.select(selectVehicleTechRecordModelHavingStatusAll)
      .pipe(
        map(s => s[0].vin)
      )),
    switchMap(([action, vin]) => this._technicalRecordService.getDocumentBlob(vin, action.filename)
      .pipe(
        switchMap((response: { buffer: ArrayBuffer, contentType: string, fileName?: string }) => {
          this._store$.dispatch(new ClearErrorMessage())
          const fileblob = new Blob([response.buffer], { type: response.contentType });
          this._FileSaverService.save(fileblob, response.fileName);
          return of(new DownloadDocumentFileActionSuccess({ blob: fileblob, fileName: response.fileName }));
        }),
        catchError((error) => {
          this._store$.dispatch(new SetErrorMessage([error.error]));
          return of(new DownloadDocumentFileActionFailure(error))
        }
        ))));

  constructor(
    private _actions$: Actions,
    private _technicalRecordService: TechnicalRecordService,
    private _store$: Store<IAppState>,
    private _FileSaverService: FileSaverService) { }
}
