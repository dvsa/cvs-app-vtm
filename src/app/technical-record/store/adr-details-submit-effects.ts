import { Injectable } from '@angular/core';
import { Actions, Effect, EffectNotification, ofType, OnRunEffects } from '@ngrx/effects';
import { Observable, of } from 'rxjs';
import { tap, exhaustMap, switchMap, map, catchError, withLatestFrom, take } from 'rxjs/operators';
import { TechnicalRecordService } from '@app/technical-record-search/technical-record.service';
import { Store } from '@ngrx/store';
import { Router } from '@angular/router';
import { selectVehicleTechRecordModelHavingStatusAll } from '@app/store/selectors/VehicleTechRecordModel.selectors';
import { SubmitAdrAction, SubmitAdrActionFailure, SubmitAdrActionSuccess } from './adrDetailsSubmit.actions';
import { GetVehicleTechRecordModelHavingStatusAllSuccess } from '@app/store/actions/VehicleTechRecordModel.actions';
import { GetVehicleTestResultModel } from '@app/store/actions/VehicleTestResultModel.actions';
import { IAppState } from '@app/store/state/app.state';
import { SetValueAction, ResetAction } from 'ngrx-forms';
import { INITIAL_STATE } from '@app/technical-record/adr-details-form/store/adrDetailsForm.state';
import { IAppState as IAdrState } from '@app/technical-record/adr-details-form/store/adrDetailsForm.state'
import { FilterRecordPipe } from '@app/pipes/FilterRecordPipe';

@Injectable()
export class AdrDetailsSubmitEffects implements OnRunEffects {
  @Effect()
  adrSubmit$ = this._actions$.pipe(
    ofType<SubmitAdrAction>(SubmitAdrAction.TYPE),
    withLatestFrom(this._store$.select(selectVehicleTechRecordModelHavingStatusAll)
      .pipe(
        map(s => {
          return {
            vin: s.vin,
            activeRecord: this._filterRecordPipe.transform(s.techRecord)
          };
        }),
      )),
    switchMap(([action, payload]) => {
      return this._store$.select((s: IAdrState) => s.adrDetails.formState).pipe(
        take(1),
        tap((_) => {
          console.log(`payload of this._store$.select((s: IAdrState) => ) => ${JSON.stringify(_)}`)
        }),
        map(fs => this.createSubmitState(fs.value, payload.activeRecord)),
      ).pipe(
        switchMap((submitValue) => this._technicalRecordService.updateTechnicalRecords(submitValue, payload.vin)
          .pipe(
            switchMap((searchIdentifier: string) => this._technicalRecordService.getTechnicalRecordsAllStatuses(payload.vin)
              .pipe(
                switchMap((techRecordJson: any) => of(new GetVehicleTechRecordModelHavingStatusAllSuccess(techRecordJson))),
                tap((_) => {
                  this._store$.dispatch(new SetValueAction(INITIAL_STATE.id, INITIAL_STATE.value));
                  this._store$.dispatch(new ResetAction(INITIAL_STATE.id));
                  this._store$.dispatch(new SubmitAdrActionSuccess({}));
                  this._store$.dispatch(new GetVehicleTestResultModel(payload.vin));
                  this._router$.navigate([`/technical-record`]);
                })
              )),
            catchError((error) =>
              of(new SubmitAdrActionFailure(error))
            ))));
    })
  );

  ngrxOnRunEffects(resolvedEffects$: Observable<EffectNotification>) {
    return this._actions$.pipe(
      exhaustMap((res) => {
        return resolvedEffects$;
      })
    );
  }

  constructor(
    private _actions$: Actions,
    private _technicalRecordService: TechnicalRecordService,
    private _store$: Store<IAppState>,
    private _router$: Router,
    private _filterRecordPipe: FilterRecordPipe
  ) { }

  private createSubmitState = (adrDetails: any, techRecord: any) => {
    const techRecordDTO: any = {};
    techRecord.files = [];
    techRecordDTO.msUserDetails = {'msUser': 'catalin' , 'msOid': '123243424-234234245'};
    techRecordDTO.techRecord = [];
    techRecord.reasonForCreation = 'Update ADR TYPE';
    techRecord.adrDetails = {
        vehicleDetails: {
          type: adrDetails.type,
          approvalDate: adrDetails.approvalDate.year + '-' + adrDetails.approvalDate.month + '-' + adrDetails.approvalDate.day
        },
        listStatementApplicable: adrDetails.listStatementApplicable === 'applicable',
        batteryListNumber: adrDetails.batteryListNumber,
        declarationSeen: adrDetails.declarationsSeen === 'true', // boolean
        brakeDeclarationsSeen: adrDetails.brakeDeclarationsSeen,
        brakeDeclarationIssuer: adrDetails.brakeDeclarationIssuer,
        brakeEndurance: adrDetails.brakeEndurance, // boolean
        weight: adrDetails.weight,
        compatibilityGroupJ: adrDetails.compatibilityGroupJ,
        permittedDangerousGoods: [
          adrDetails.permittedDangerousGoods
        ],
        additionalExaminerNotes: adrDetails.additionalNotes,
        applicantDetails: {
          name: adrDetails.name,
          street: adrDetails.street,
          town: adrDetails.town,
          city: adrDetails.city,
          postcode: adrDetails.postcode
        },
        memosApply: [
          adrDetails.memosApply
        ],
        additionalNotes: {
          guidanceNotes: [
            adrDetails.guidanceNotes
          ]
        },
        adrTypeApprovalNo: adrDetails.adrTypeApprovalNo,
        tank: {
          tankDetails: {
            tankManufacturer: adrDetails.tankManufacturer,
            yearOfManufacture: adrDetails.yearOfManufacture,
            tankCode: adrDetails.tankCode,
            specialProvisions: adrDetails.specialProvisions,
            tankManufacturerSerialNo: adrDetails.tankManufacturerSerialNo,
            tankTypeAppNo: adrDetails.tankTypeAppNo,
            tc2Details: {
              tc2Type: adrDetails.tc2Type,
              tc2IntermediateApprovalNo: adrDetails.tc2IntermediateApprovalNo,
              // tslint:disable-next-line:max-line-length
              tc2IntermediateExpiryDate: adrDetails.tc2IntermediateExpiryDate.year + '-' + adrDetails.tc2IntermediateExpiryDate.month + '-' + adrDetails.tc2IntermediateExpiryDate.day
            },
            tc3Details: [
              {
                tc3Type: adrDetails.tc3Type,
                tc3PeriodicNumber: adrDetails.tc3PeriodicNumber,
                // tslint:disable-next-line:max-line-length
                tc3PeriodicExpiryDate: adrDetails.tc3PeriodicExpiryDate.year + '-' + adrDetails.tc3PeriodicExpiryDate.month + '-' + adrDetails.tc3PeriodicExpiryDate.day
              }
            ]

          },
          tankStatement: {
            substancesPermitted: adrDetails.substancesPermitted,
            statement: adrDetails.statement,
            productListRefNo: adrDetails.productListRefNo,
            productListUnNo: adrDetails.productListUnNo,
            productList: adrDetails.productList
          }
        }
      };

      techRecordDTO.techRecord.push(techRecord);
      return techRecordDTO;
  }
}
