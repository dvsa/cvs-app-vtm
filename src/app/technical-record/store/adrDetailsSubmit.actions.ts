import { Action } from '@ngrx/store';

export class SubmitAdrAction implements Action {
    static readonly TYPE = 'adrDetails/SUBMIT_ADR_ACTION';
    readonly type = SubmitAdrAction.TYPE;
    constructor(public submitContext: string) { }
}

export class SubmitAdrActionSuccess implements Action {
    static readonly TYPE = 'adrDetails/SUBMIT_ADR_ACTION_SUCCESS';
    readonly type = SubmitAdrActionSuccess.TYPE;
    constructor(public payload: any) { }
}

export class SubmitAdrActionFailure implements Action {
    static readonly TYPE = 'adrDetails/SUBMIT_ADR_ACTION_FAILURE';
    readonly type = SubmitAdrActionFailure.TYPE;
    constructor(public payload: any) { }
}