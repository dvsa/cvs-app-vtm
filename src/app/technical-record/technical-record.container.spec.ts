import { Component, Input, EventEmitter, Output, DebugElement } from '@angular/core';
import { RouterTestingModule } from '@angular/router/testing';
import { SpyLocation } from '@angular/common/testing';
import { By } from '@angular/platform-browser';
import { ComponentFixture, TestBed } from '@angular/core/testing';
import { Store, Action } from '@ngrx/store';
import { BehaviorSubject, Observable } from 'rxjs';
import { map } from 'rxjs/operators';

import { TESTING_UTILS } from './../utils/testing.utils';
import { SharedModule } from '@app/shared';
import {
  getSelectedVehicleTechRecord,
  getVehicleTechRecordMetaData,
  getViewState,
  getTechRecord
} from '@app/store/selectors/VehicleTechRecordModel.selectors';
import {
  UpdateVehicleTechRecord,
  SetViewState
} from '@app/store/actions/VehicleTechRecordModel.actions';
import { TechRecord } from '@app/models/tech-record.model';
import { VehicleTechRecordModel } from '@app/models/vehicle-tech-record.model';
import { MetaData } from '@app/models/meta-data';
import { VIEW_STATE } from '@app/app.enums';
import { TechnicalRecordsContainer } from './technical-record.container';
import { TestResultModel } from '@app/models/test-result.model';

const mockSelector = new BehaviorSubject<any>(undefined);

class MockStore {
  select(selector: any) {
    switch (selector) {
      case getSelectedVehicleTechRecord:
        return mockSelector.pipe(
          map((value: any) =>
            value && value.hasOwnProperty('getSelectedVehicleTechRecord')
              ? value['getSelectedVehicleTechRecord']
              : {}
          )
        );

      case getTechRecord:
        return mockSelector.pipe(
          map((value: any) =>
            value && value.hasOwnProperty('getTechRecord') ? value['getTechRecord'] : {}
          )
        );

      case getVehicleTechRecordMetaData:
        return mockSelector.pipe(
          map((value: any) =>
            value && value.hasOwnProperty('getVehicleTechRecordMetaData')
              ? value['getVehicleTechRecordMetaData']
              : {}
          )
        );

      case getViewState:
        return mockSelector.pipe(
          map((value: any) =>
            value && value.hasOwnProperty('getViewState') ? value['getViewState'] : {}
          )
        );
      default:
        return mockSelector;
    }
  }

  dispatch(action: Action) {
    return [];
  }
}

describe('TechnicalRecordsContainer', () => {
  let fixture: ComponentFixture<TechnicalRecordsContainer>;
  let component: TechnicalRecordsContainer;
  const store: MockStore = new MockStore();

  beforeEach(() => {
    TestBed.configureTestingModule({
      imports: [SharedModule, RouterTestingModule],
      declarations: [TechnicalRecordsContainer, TestTechnicalRecordsComponent],
      providers: [
        { provide: Store, useValue: store },
        { provide: Location, useClass: SpyLocation }
      ]
    }).compileComponents();

    fixture = TestBed.createComponent(TechnicalRecordsContainer);
    component = fixture.componentInstance;
    jest.spyOn(store, 'dispatch');

    fixture.detectChanges();
  });

  it('should render the container', () => {
    mockSelector.next({
      getSelectedVehicleTechRecord: {
        techRecord: [TESTING_UTILS.mockTechRecord()]
      } as VehicleTechRecordModel,
      getTechRecord: TESTING_UTILS.mockTechRecord(),
      getVehicleTechRecordMetaData: TESTING_UTILS.mockMetaData(),
      getViewState: VIEW_STATE.VIEW_ONLY
    });

    fixture.detectChanges();

    expect(fixture).toMatchSnapshot();
  });

  describe('action', () => {
    let mockTechComponent: DebugElement;
    beforeEach(() => {
      mockTechComponent = fixture.debugElement.query(By.directive(TestTechnicalRecordsComponent));
    });

    it('should dispatch for the edited techRecord', () => {
      const editedTechRecord = { statusCode: 'current' } as TechRecord;

      const updateVehicleTechRecordAction = new UpdateVehicleTechRecord(editedTechRecord);
      mockTechComponent.componentInstance.submitTechRecord.emit(editedTechRecord);

      expect(store.dispatch).toHaveBeenCalledWith(updateVehicleTechRecordAction);
    });

    it('should dispatch for the view state', () => {
      const viewStateAction = new SetViewState(VIEW_STATE.EDIT);
      mockTechComponent.componentInstance.changeViewState.emit(VIEW_STATE.EDIT);

      expect(store.dispatch).toHaveBeenCalledWith(viewStateAction);
    });
  });
});

@Component({
  selector: 'vtm-technical-record',
  template: `
    <div>Current or Selected Vehicle Tech Record: {{ vehicleTechRecord | json }}</div>
    <div>Active record based on Statuscode: {{ activeRecord | json }}</div>
    <div>MetaData: {{ metaData | json }}</div>
    <div>View state is: {{ editState ? 'EDIT' : 'VIEW ONLY' }}</div>
  `
})
class TestTechnicalRecordsComponent {
  @Input() vehicleTechRecord: VehicleTechRecordModel;
  @Input() activeRecord: TechRecord;
  @Input() metaData: MetaData;
  @Input() testResultJson: TestResultModel;
  @Input() editState: VIEW_STATE;
  @Input() canEdit: Observable<boolean>;
  @Output() submitTechRecord = new EventEmitter<TechRecord>();
  @Output() changeViewState = new EventEmitter<VIEW_STATE>();
}
