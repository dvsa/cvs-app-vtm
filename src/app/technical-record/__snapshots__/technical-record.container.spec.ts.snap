// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TechnicalRecordsContainer should render the container 1`] = `
<vtm-technical-record-container
  activeTechRecord$={[Function AnonymousSubject]}
  canEdit$={[Function Observable]}
  createState$={[Function Observable]}
  metaData$={[Function AnonymousSubject]}
  store={[Function MockStore]}
  testResults$={[Function BehaviorSubject]}
  vehicleTechnicalRecord$={[Function AnonymousSubject]}
  viewState$={[Function AnonymousSubject]}
>
  <div
    class="govuk-width-container"
  >
    <a
      class="govuk-back-link"
      vtmbackbutton=""
    >
      Back
    </a>
    <main
      class="govuk-main-wrapper"
    >
      <vtm-technical-record
        ng-reflect-can-edit="0"
        ng-reflect-edit-state="0"
        ng-reflect-meta-data="[object Object]"
        ng-reflect-test-result-json="[object Object]"
        ng-reflect-vehicle-tech-record="[object Object]"
      >
        <div>
          Current or Selected Vehicle Tech Record: {
    "techRecord": [
      {
        "statusCode": "provisional",
        "adrDetails": {
          "vehicleDetails": {
            "approvalDate": "2020-03-10",
            "type": "centre axle battery"
          },
          "listStatementApplicable": false,
          "batteryListNumber": "",
          "declarationsSeen": false,
          "brakeDeclarationsSeen": false,
          "brakeDeclarationIssuer": "",
          "brakeEndurance": false,
          "weight": "",
          "compatibilityGroupJ": false,
          "documents": [],
          "permittedDangerousGoods": [
            "AT"
          ],
          "additionalExaminerNotes": "",
          "applicantDetails": {
            "name": "Ben",
            "street": "Robert green",
            "city": "Birmingham",
            "town": "lala land",
            "postcode": "NG4 12Z"
          },
          "memosApply": [],
          "additionalNotes": {
            "number": [],
            "guidanceNotes": []
          },
          "adrTypeApprovalNo": ""
        }
      }
    ]
  }
        </div>
        <div>
          Active record based on Statuscode: null
        </div>
        <div>
          MetaData: {
    "adrDetails": {
      "memosApplyFe": [
        "07/09 3mth leak ext"
      ],
      "tank": {
        "tankStatement": {
          "substancesPermittedFe": [
            "Substances permitted under the tank code and any special provisions specified in 9 may be carried",
            "Substances (Class UN number and if necessary packing group and proper shipping name) may be carried"
          ]
        }
      },
      "additionalNotes": {
        "guidanceNotesFe": [
          "New certificate requested"
        ],
        "numberFe": [
          "V1B",
          "T1B"
        ]
      },
      "permittedDangerousGoodsFe": [
        "FP &lt;61 (FL)",
        "AT",
        "Explosives (type 2)"
      ],
      "vehicleDetails": {
        "typeFe": [
          "Artic tractor",
          "Semi trailer battery",
          "Centre axle tank",
          "Centre axle battery"
        ]
      }
    }
  }
        </div>
        <div>
          View state is: VIEW ONLY
        </div>
      </vtm-technical-record>
    </main>
  </div>
</vtm-technical-record-container>
`;
